src/core/display/Container.js

```diff
@@ -428,59 +428,85 @@ Container.prototype.renderWebGL = function (renderer)
         return;
     }
 
-    var i, j;
 
     // do a quick check to see if this element has a mask or a filter.
     if (this._mask || this._filters)
     {
-        renderer.currentRenderer.flush();
-
-        // push filter first as we need to ensure the stencil buffer is correct for any masking
-        if (this._filters && this._filters.length)
-        {
-            renderer.filterManager.pushFilter(this, this._filters);
-        }
+        this.renderAdvancedWebGL(renderer);
+    }
+    else
+    {
+        this._renderWebGL(renderer);
 
-        if (this._mask)
+        // simple render children!
+        for (var i = 0, j = this.children.length; i < j; ++i)
         {
-            renderer.maskManager.pushMask(this, this._mask);
+            this.children[i].renderWebGL(renderer);
         }
+    }
+};
 
-        renderer.currentRenderer.start();
+Container.prototype.renderAdvancedWebGL = function (renderer)
+{
+    renderer.currentRenderer.flush();
 
-        // add this object to the batch, only rendered if it has a texture.
-        this._renderWebGL(renderer);
+    var filters = this._filters;
+    var mask = this._mask;
+    var i, j;
 
-        // now loop through the children and make sure they get rendered
-        for (i = 0, j = this.children.length; i < j; i++)
+    // push filter first as we need to ensure the stencil buffer is correct for any masking
+    if ( filters )
+    {
+        if(!this._enabledFilters)
         {
-            this.children[i].renderWebGL(renderer);
+            this._enabledFilters = [];
         }
 
-        renderer.currentRenderer.flush();
+        this._enabledFilters.length = 0;
 
-        if (this._mask)
+        for (i = 0; i < filters.length; i++)
         {
-            renderer.maskManager.popMask(this, this._mask);
+            if(filters[i].enabled)
+            {
+                this._enabledFilters.push( filters[i] );
```

Comment:
Good question! Even though it's cleaner to look at , it generates a new
array each time which is something I'm keen to avoid in the core render
loop..
On Sat, 13 Aug 2016 at 01:04, Matt Karl <notifications@github.com> wrote:

> In src/core/display/Container.js
> <https://github.com/pixijs/pixi.js/pull/2798#discussion_r74676003>:
>
> >          {
> > -            renderer.maskManager.popMask(this, this._mask);
> > +            if(filters[i].enabled)
> > +            {
> > +                this._enabledFilters.push( filters[i] );
>
> Why not use Array.filter?
>
> this._enableFilters = filters.filter(function(filter){ return
> filter.enabled;});
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pixijs/pixi.js/pull/2798/files/3d10d48c974de937316c9b3f50134282402819bb#r74676003>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC998rzT-wrkpGR3C7zwsMj81WIdjOT-ks5qfQn_gaJpZM4Jjcqt>
> .
>